ipset.go

```diff
@@ -99,154 +99,135 @@ func (s *IPSet) RemoveSet(b *IPSet) {
-// See also comments in netaddr_test.go's TestPointLess.
-func (a point) Less(b point) bool {
-	cmp := a.ip.Compare(b.ip)
-	if cmp != 0 {
-		return cmp < 0
-	}
-	if a.want != b.want {
-		if a.start == b.start {
-			return !a.want
-		}
-		return a.start
-	}
-	if a.start != b.start {
-		return a.start
-	}
-	return false
-}
-
 func discardf(format string, args ...interface{}) {}
 
 // debugf is reassigned by tests.
 var debugf = discardf
 
-func debugLogPoints(points []point) {
-	for _, p := range points {
-		emo := "✅"
-		if !p.want {
-			emo = "❌"
-		}
-		if p.start {
-			debugf(" {  %-15s %s\n", p.ip, emo)
-		} else {
-			debugf("  } %-15s %s\n", p.ip, emo)
-		}
-	}
-}
-
 // Ranges returns the minimum and sorted set of IP
 // ranges that covers s.
 func (s *IPSet) Ranges() []IPRange {
-	var points []point
-	for _, r := range s.in {
-		points = append(points, point{r.From, true, true}, point{r.To, true, false})
+	const debug = false
+	if debug {
+		debugf("ranges start in=%v out=%v", s.in, s.out)
 	}
-	for _, r := range s.out {
-		points = append(points, point{r.From, false, true}, point{r.To, false, false})
+	in, ok := mergeIPRanges(s.in)
+	if !ok {
+		return nil
+	}
+	out, ok := mergeIPRanges(s.out)
+	if !ok {
+		return nil
 	}
-	sort.Slice(points, func(i, j int) bool { return points[i].Less(points[j]) })
-	const debug = false
 	if debug {
-		debugf("post-sort:")
-		debugLogPoints(points)
-		debugf("merging...")
+		debugf("ranges sort  in=%v out=%v", in, out)
 	}
 
-	// Now build 'want', like points but with "remove" ranges removed
-	// and adjacent blocks merged, and all elements alternating between
-	// start and end.
-	want := points[:0]
-	var addDepth, removeDepth int
-	for i, p := range points {
-		depth := &addDepth
-		if !p.want {
-			depth = &removeDepth
-		}
-		if p.start {
-			*depth++
-		} else {
-			*depth--
-		}
+	// in and out are sorted in ascending range order, and have no
+	// overlaps within each other. We can run a merge of the two lists
+	// in one pass.
+
+	ret := make([]IPRange, 0, len(in))
+	for len(in) > 0 && len(out) > 0 {
 		if debug {
-			debugf("at[%d] (%+v), add=%v, remove=%v", i, p, addDepth, removeDepth)
-		}
-		if p.start && *depth != 1 {
-			continue
-		}
-		if !p.start && *depth != 0 {
-			continue
+			debugf("step in=%v out=%v", in[0], out[0])
 		}
-		if !p.want && addDepth > 0 {
-			if p.start {
-				// If we're transitioning from a range of
-				// addresses we want to ones we don't, insert
-				// an end marker for the IP before the one we
-				// don't.
-				want = append(want, point{
-					ip:    p.ip.Prior(),
-					want:  true,
-					start: false,
-				})
-			} else {
-				want = append(want, point{
-					ip:    p.ip.Next(),
-					want:  true,
-					start: true,
-				})
+
+		switch {
+		case !out[0].Valid() || !in[0].Valid():
+			// mergeIPRanges should have prevented invalid ranges from
+			// sneaking in.
+			panic("invalid IPRanges during Ranges merge")
+		case out[0].To.Less(in[0].From):
```

Comment:
seems like most of these cases could/should be unexported methods on IPRange would let you move docs to the methods, and let them be tested easily separately too.

```go
   case outr.entirelyBefore(inr):
```

(or out[0] if you don't like names for them)